### ISSUE TITLE
Build fails with ReferenceError("navigator is not defined")

### ISSUE THREAD

**Issue by alfaburro** [fastly/remix-compute-js#8]
Hi, I've been following the instructions in [this](https://www.fastly.com/blog/host-your-remix-app-on-fastly-compute-edge) blog article and when running this command `npm run dev` the output indicates an error. What did I do wrong here?

```
gitpod /workspace/remix (main) $ npm run dev

> dev
> run-p "dev:*"


> dev:remix
> remix watch


> dev:fastly
> fastly compute serve --watch

✓ Verifying fastly.toml
✓ Identifying package name
✓ Identifying toolchain
- Running [scripts.build]...Watching Remix app in development mode...
\ Running [scripts.build]...⚠️ REMIX FUTURE CHANGE: The behaviors of `CatchBoundary` and `ErrorBoundary` are changing in v2. You can prepare for this change at your convenience with the `v2_errorBoundary` future flag. For instructions on making this change see https://remix.run/docs/en/v1.15.0/pages/v2#catchboundary-and-errorboundary
⚠️ REMIX FUTURE CHANGE: APIs that provide `formMethod` will be changing in v2. All values will be uppercase (GET, POST, etc.) instead of lowercase (get, post, etc.) You can prepare for this change at your convenience with the `v2_normalizeFormMethod` future flag. For instructions on making this change see https://remix.run/docs/en/v1.15.0/pages/v2#formMethod
⚠️ REMIX FUTURE CHANGE: The route `meta` export signature is changing in v2. You can prepare for this change at your convenience with the `v2_meta` future flag. For instructions on making this change see https://remix.run/docs/en/v1.15.0/pages/v2#meta
⚠️ REMIX FUTURE CHANGE: The route file convention is changing in v2. You can prepare for this change at your convenience with the `v2_routeConvention` future flag. For instructions on making this change see https://remix.run/docs/en/v1.15.0/pages/v2#file-system-route-convention
- Running [scripts.build]...💿 Building...
- Running [scripts.build]...💿 Rebuilt in 1.1s
✗ Running [scripts.build]

INFO: Command output:
--------------------------------------------------------------------------------

> build
> npm run build:remix && npm run build:fastly


> build:remix
> remix build

Building Remix app in production mode...
⚠️ REMIX FUTURE CHANGE: The behaviors of `CatchBoundary` and `ErrorBoundary` are changing in v2. You can prepare for this change at your convenience with the `v2_errorBoundary` future flag. For instructions on making this change see https://remix.run/docs/en/v1.15.0/pages/v2#catchboundary-and-errorboundary
⚠️ REMIX FUTURE CHANGE: APIs that provide `formMethod` will be changing in v2. All values will be uppercase (GET, POST, etc.) instead of lowercase (get, post, etc.) You can prepare for this change at your convenience with the `v2_normalizeFormMethod` future flag. For instructions on making this change see https://remix.run/docs/en/v1.15.0/pages/v2#formMethod
⚠️ REMIX FUTURE CHANGE: The route `meta` export signature is changing in v2. You can prepare for this change at your convenience with the `v2_meta` future flag. For instructions on making this change see https://remix.run/docs/en/v1.15.0/pages/v2#meta
⚠️ REMIX FUTURE CHANGE: The route file convention is changing in v2. You can prepare for this change at your convenience with the `v2_routeConvention` future flag. For instructions on making this change see https://remix.run/docs/en/v1.15.0/pages/v2#file-system-route-convention
Built in 545ms

> prebuild:fastly
> compute-js-static-publish --build-static --suppress-framework-warnings && webpack

Fastly Compute@Edge JavaScript Static Publisher
🚀 Building loader...
✔️ Public directory '/workspace/remix'.
✔️ Using static directories: ./public/build
✔️ Using exclude directories: ./node_modules
✔️ Applying 0 custom content type(s).
✔️ Using include directories: ./build, ./public
✔️ /build/index.js: "application/javascript"
✔️ Copied static string file "/workspace/remix/build/index.js" to "./src/static-content/file0.txt".
✔️ /build/index.js.map: "application/json"
✔️ Copied static string file "/workspace/remix/build/index.js.map" to "./src/static-content/file1.txt".
✔️ /public/build/_assets/index-4CHP7APY.css: "text/css" [STATIC]
✔️ Copied static string file "/workspace/remix/public/build/_assets/index-4CHP7APY.css" to "./src/static-content/file2.txt".
✔️ /public/build/_shared/chunk-5XCBIPQZ.js: "application/javascript" [STATIC]
✔️ Copied static string file "/workspace/remix/public/build/_shared/chunk-5XCBIPQZ.js" to "./src/static-content/file3.txt".
✔️ /public/build/_shared/chunk-DQFVS2MO.js: "application/javascript" [STATIC]
✔️ Copied static string file "/workspace/remix/public/build/_shared/chunk-DQFVS2MO.js" to "./src/static-content/file4.txt".
✔️ /public/build/entry.client-4M6NKTJX.js: "application/javascript" [STATIC]
✔️ Copied static string file "/workspace/remix/public/build/entry.client-4M6NKTJX.js" to "./src/static-content/file5.txt".
✔️ /public/build/manifest-371E4DDA.js: "application/javascript" [STATIC]
✔️ Copied static string file "/workspace/remix/public/build/manifest-371E4DDA.js" to "./src/static-content/file6.txt".
✔️ /public/build/root-RBQKXDDT.js: "application/javascript" [STATIC]
✔️ Copied static string file "/workspace/remix/public/build/root-RBQKXDDT.js" to "./src/static-content/file7.txt".
✔️ /public/build/routes/index-VEUIAML6.js: "application/javascript" [STATIC]
✔️ Copied static string file "/workspace/remix/public/build/routes/index-VEUIAML6.js" to "./src/static-content/file8.txt".
✔️ /public/favicon.ico: "image/vnd.microsoft.icon"
✔️ Copied static binary file "/workspace/remix/public/favicon.ico" to "./src/static-content/file9.bin".
✔️ /remix.config.js: "application/javascript"
✔️ Copied static string file "/workspace/remix/remix.config.js" to "./src/static-content/file10.txt".
✅  Wrote static file loader for 11 file(s).
✅  Wrote static file loader types file statics.d.ts.
asset index.js 816 KiB [compared for emit] [minimized] [big] (name: main) 1 related asset
orphan modules 416 KiB [orphan] 30 modules
runtime modules 1010 bytes 5 modules
modules by path ./node_modules/ 1.32 MiB
  javascript modules 1.3 MiB 157 modules
  json modules 12.7 KiB 6 modules
optional modules 280 bytes [optional]
  ./build/ lazy namespace object 160 bytes [optional] [built] [code generated]
  buffer (ignored) 15 bytes [optional] [built] [code generated]
  buffer (ignored) 15 bytes [optional] [built] [code generated]
  + 6 modules
./src/index.js + 4 modules 213 KiB [not cacheable] [built] [code generated]
./remix.config.js 474 bytes [built] [code generated]
util (ignored) 15 bytes [built] [code generated]
util (ignored) 15 bytes [built] [code generated]

WARNING in ./build/index.js 19:221-237
Critical dependency: the request of a dependency is an expression
 @ ./src/statics.js 7:0-49 12:163-174
 @ ./src/index.js 3:0-41 5:47-59

WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value.
Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/

WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  index.js (816 KiB)

WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  main (816 KiB)
      index.js

WARNING in webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/

1 warning has detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

webpack 5.85.0 compiled with 5 warnings in 8448 ms

> build:fastly
> js-compute-runtime ./bin/index.js ./bin/main.wasm

Exception while evaluating JS: (new ReferenceError("navigator is not defined", "<stdin>", 14950))
  bin/index.js/k<@<stdin>:14950:16
  bin/index.js/f/<@<stdin>:14928:229
  bin/index.js/x<@<stdin>:14955:13
  bin/index.js/f/<@<stdin>:14928:229
  bin/index.js/O<@<stdin>:15863:33
  bin/index.js/f/<@<stdin>:14928:229
  bin/index.js/D<@<stdin>:15865:13
  bin/index.js/f/<@<stdin>:14928:229
  bin/index.js/<@<stdin>:20368:26
  bin/index.js/<@<stdin>:20837:11
  bin/index.js@<stdin>:20840:9
  __require@<stdin>:116:52
  @<stdin>:20843:3
  @<stdin>:20844:3

Error: the `wizer.initialize` function trapped

Caused by:
    Exited with i32 exit status 1
    wasm backtrace:
        0: 0x9d910b - <unknown>!<wasm function 21439>
        1: 0x9d8e23 - <unknown>!<wasm function 21415>
        2: 0x9da803 - <unknown>!<wasm function 21467>
        3: 0x11504 - <unknown>!<wasm function 225>
        4: 0x11bfe - <unknown>!<wasm function 228>
        5: 0x11df2 - <unknown>!<wasm function 230>
    
Error: Failed to compile JavaScript to Wasm:  Wizer initialization failure
--------------------------------------------------------------------------------

ERROR: error during execution process (see 'command output' above): exit status 1.

Here are some steps you can follow to debug the issue:

- Re-run the fastly command with the --verbose flag to see more information.
- Is the required language toolchain (node/npm, rust/cargo etc) installed correctly?
- Is the required version (if any) of the language toolchain installed/activated?
- Were the required dependencies (package.json, Cargo.toml etc) installed?
- Did the build script (see fastly.toml [scripts.build]) produce a ./bin/main.wasm binary file?
- Was there a configured [scripts.post_build] step that needs to be double-checked?

For more information on fastly.toml configuration settings, refer to https://developer.fastly.com/reference/compute/fastly-toml/
ERROR: "dev:fastly" exited with 1.
gitpod /workspace/remix (main) $ 
```

**Comment by harmony7** [comment]
Hi, thanks for reaching out!
Will be investigating this.

**Comment by harmony7** [comment]
Thanks for looking into this.
You said in your comment that you upgraded the project for compute-js-static-publish@4, and removed Webpack.

Unfortunately, the current version of the adapter is incompatible with compute-js-static-publish@4, and in addition, it relies on polyfilling `crypto` by using the `crypto-browserify` module and using Webpack to bundle it. I'm working on an update to support v4, but for now you'll have to use v3 and keep Webpack.

I dug into the root cause of this, and it looks like this is actually a bug in Remix that was introduced in v1.16, still broken in v1.17.
See:
https://github.com/remix-run/remix/issues/6348

However, Remix has a fix and it has already been merged (was merged just a few hours ago!). So we should see it working again in their next release.
See:
https://github.com/remix-run/remix/pull/6562

**Comment by MichaelDeBoey** [comment]
We just published version `1.17.1-pre.0` which includes https://github.com/remix-run/remix/pull/6562 (a possible fix). If you'd like to take it for a test run please try it out and let us know what you think!

**Comment by harmony7** [comment]
@MichaelDeBoey Thanks! I've tried it out and it works, the error above is now cleared up.

@alfaburro Can you try again, using this version of the package?
Revert your changes, and then install 1.17.1-pre.0:

```
npm install @remix-run/react@1.17.1-pre.0
npm install --save-dev @remix-run/dev@1.17.1-pre.0 @remix-run/eslint-config@1.17.1-pre.0
```

Now, `npm run dev` should work without the `navigator is not defined` error. If this works for you then that will be good feedback for @MichaelDeBoey and the Remix team as well.

**Comment by harmony7** [comment]
If you're re-scaffolding your app, you can use the `--remix-version` command line option to specify that you want to use this `1.17.1-pre.0` version of Remix:

```
% npx create-remix@latest ./my-app --template https://github.com/fastly/remix-compute-js/tree/main/packages/remix-template --remix-version 1.17.1-pre.0
```

Or, try out the new v2 beta, discussed here: https://github.com/fastly/remix-compute-js/issues/9. This version uses the updated compute-js-static-publish@5.

```
% npx create-remix@latest ./my-app --template https://github.com/fastly/remix-compute-js/tree/v2.0.0-beta.0/packages/remix-template --remix-version 1.17.1-pre.0
```

I'm closing this out, as we know the problem is with remix 1.16.x/1.17.0, and that it will be fixed once 1.17.1 is released. But please feel free to reopen if you have questions or need clarification.

**Comment by alfaburro** [comment]
Looks good now, thank you for looking into this @harmony7
